CICE/patches/ice_domain.F90.patch

```diff
@@ -0,0 +1,71 @@
+diff --git a/cicecore/cicedyn/infrastructure/ice_domain.F90 b/cicecore/cicedyn/infrastructure/ice_domain.F90
+index 8b680f2..cf13fcc 100644
+--- a/cicecore/cicedyn/infrastructure/ice_domain.F90
++++ b/cicecore/cicedyn/infrastructure/ice_domain.F90
+@@ -88,7 +88,7 @@ module ice_domain
+        distribution_wght_file  ! file for distribution_wght=file
+ 
+     integer (int_kind) :: &
+-       nprocs                ! num of processors
++       nprocs, nprocs_x, nprocs_y                ! num of processors
+ 
+ !***********************************************************************
+ 
+@@ -101,7 +101,7 @@ subroutine init_domain_blocks
+ !  This routine reads in domain information and calls the routine
+ !  to set up the block decomposition.
+ 
+-   use ice_distribution, only: processor_shape
++   use ice_distribution, only: processor_shape, proc_decomposition
+    use ice_domain_size, only: ncat, nilyr, nslyr, max_blocks, &
+        nx_global, ny_global, block_size_x, block_size_y
+    use ice_fileunits, only: goto_nml
+@@ -204,6 +204,14 @@ subroutine init_domain_blocks
+ 
+    endif
+ 
++   if (nprocs .ne. -1) then
++      if (my_task == master_task) then
++         write(nu_diag,*) subname//' WARNING: nprocs is deprecated, please remove from namelist'
++      endif
++   endif
++
++   nprocs = get_num_procs()
++
+    call broadcast_scalar(nprocs,            master_task)
+    call broadcast_scalar(processor_shape,   master_task)
+    call broadcast_scalar(distribution_type, master_task)
+@@ -216,10 +224,14 @@ subroutine init_domain_blocks
+    call broadcast_scalar(maskhalo_bound,    master_task)
+    call broadcast_scalar(add_mpi_barriers,  master_task)
+    call broadcast_scalar(debug_blocks,      master_task)
++
++   ! update nprocs_x and nprocs_y
```

Comment:
I would write this in a slightly different way:

```fortran
  ! Broadcast data from namelist
   call broadcast_scalar(processor_shape,   master_task)
   call broadcast_scalar(distribution_type, master_task)
   call broadcast_scalar(distribution_wght, master_task)
   call broadcast_scalar(distribution_wght_file, master_task)
   call broadcast_scalar(ew_boundary_type,  master_task)
   call broadcast_scalar(ns_boundary_type,  master_task)
   call broadcast_scalar(maskhalo_dyn,      master_task)
   call broadcast_scalar(maskhalo_remap,    master_task)
   call broadcast_scalar(maskhalo_bound,    master_task)
   call broadcast_scalar(add_mpi_barriers,  master_task)
   call broadcast_scalar(debug_blocks,      master_task)
   call broadcast_scalar(max_blocks,      master_task)
   call broadcast_scalar(block_size_x,      master_task)
   call broadcast_scalar(block_size_y,      master_task)
   call broadcast_scalar(nx_global,         master_task)
   call broadcast_scalar(ny_global,         master_task) 

   nprocs = get_num_procs()

   ! update nprocs_x and nprocs_y
   call proc_decomposition(nprocs, nprocs_x, nprocs_y)

   ! Automatically determine max_blocks if requested by user
   if (max_blocks < 1) then
      max_blocks=((nx_global-1)/block_size_x/nprocs_x+1) * &
                 ((ny_global-1)/block_size_y/nprocs_y+1)
      max_blocks=max(1,max_blocks)
      if (my_task == master_task) then
         write(nu_diag,'(/,a52,i6,/)') &
            '(ice_domain): max_block < 1: max_block estimated to ',max_blocks
      endif
   endif
```

I think this is easier to read, but in the end this a personal preference, so feel free to ignore this suggestion.

Reply:
Thanks Micael

So `broadcast_scalar` calls will wait for that variable to be populated before continuing ?

i.e. On PEs which are not master_task, will this will wait for block_size and max_blocks to be read from the namelist before the max_blocks calculation starts?

Reply:
> i.e. On PEs which are not master_task, will this will wait for block_size and max_blocks to be read from the namelist before the max_blocks calculation starts?

Yes, exactly. All task will only return from the call to `broadcast_scalar` and continue execution once they have a local copy of the value being broadcasted.